### PR TITLE
[SPARK-39281][SQL] Speed up Timestamp type inference with legacy format  in JSON/CSV data source

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
Follow up https://github.com/apache/spark/pull/36562 , performance improvement when Timestamp type inference with legacy format.

In the current implementation of CSV/JSON data source, the schema inference with legacy format relies on methods that will throw exceptions if the fields can't convert as some data types .

Throwing and catching exceptions can be slow. We can improve it by creating methods that return optional results instead.

The optimization of DefaultTimestampFormatter has been implemented in https://github.com/apache/spark/pull/36562 , this PR adds the optimization of legacy format. The basic logic is to prevent the formatter from throwing exceptions, and then use catch to determine whether the parsing is successful.

### Why are the changes needed?

Performance improvement when Timestamp type inference with legacy format.

When use JSON datasource, the speed up `67%`. CSV datasource speed also up, but not obvious.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add new test

### CODE REVIEW

**Review by MaxGekk — CHANGES_REQUESTED** [review]
Do the benchmarks `CSVBenchmark` and `JsonBenchmark` show any improvements? Could you regenerate the results `JsonBenchmark.*.txt` and `CSVBenchmark.*.txt`, please.

**Comment by Hisoka-X** [comment]
> Do the benchmarks `CSVBenchmark` and `JsonBenchmark` show any improvements? Could you regenerate the results `JsonBenchmark.*.txt` and `CSVBenchmark.*.txt`, please.

I'm doing benchmarks, but I found a problem, like @gengliangwang said in https://github.com/apache/spark/pull/36562#issuecomment-1127600354 .The benchmarks no case for `the string inputs are not valid timestamps`. The speed up only work when `string input are not valid timestamps`. I'm worry about the benchmarks can't prove anything. Can I create a PR for add benchmarks for `type inference when string input are not valid timestamps`

**Comment by MaxGekk** [comment]
@Hisoka-X Please, resolve the conflicts and rebase on the recent master.

**Comment by Hisoka-X** [comment]
> @Hisoka-X Please, resolve the conflicts and rebase on the recent master.

Ok, I will add benchmarks for this too. Please wait. Thanks!

File: sql/core/benchmarks/JsonBenchmark-results.txt [hunk 1]
```diff
@@ -4,120 +4,120 @@ Benchmark for performance of JSON parsing
...
-date strings                                                                     469            484          24          2.1         469.3       0.6X
-parse dates from Dataset[String]                                                1565           1572          11          0.6        1564.8       0.2X
-from_json(timestamp)                                                            5825           5917          88          0.2        5824.5       0.0X
-from_json(date)                                                                 3553           3574          19          0.3        3553.1       0.1X
-infer error timestamps from Dataset[String] with default format                 2590           2609          19          0.4        2589.9       0.1X
-infer error timestamps from Dataset[String] with user-provided format           2517           2551          30          0.4        2516.8       0.1X
-infer error timestamps from Dataset[String] with legacy format                  6836           6876          63          0.1        6836.1       0.0X
```
Comment (Hisoka-X) [inline sql/core/benchmarks/JsonBenchmark-results.txt:113]: @MaxGekk Hi, I updated the benchmark, the speed already up.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @Hisoka-X.

**Pull request closed without merging**